### PR TITLE
Runtime watchers

### PR DESCRIPTION
**Pull request by joshuafcole**

### CODE REVIEW

File: src/runtime/dsl.ts [hunk 1]
```diff
@@ -296,13 +303,23 @@ class DSLRecord {
...
       throw new Error("Adds and removes may only happen in the root block.");
     }
-    let record = new DSLRecord(this.__block, [], {[attributeName]: value}, this.__record);
+    values = toArray(values);
+
+    let record = new DSLRecord(this.__block, [], {}, this.__record);
     record.__output = true;
     this.__block.records.push(record);
+
+    if(typeof attributeName === "string") {
```
Comment (ibdknox) [inline src/runtime/dsl.ts]: add should not be added to the identity, so it's always dynamic

File: src/runtime/dsl.ts [hunk 2]
```diff
@@ -1096,6 +1132,10 @@ export class Program {
   blocks:DSLBlock[] = [];
   runtimeBlocks:runtime.Block[] = [];
   index:indexes.Index;
+  nodeCount = 0;
```
Comment (ibdknox) [inline src/runtime/dsl.ts]: 🗑

File: src/runtime/dsl.ts [hunk 3]
```diff
@@ -1104,22 +1144,41 @@ export class Program {
     this.index = new indexes.HashIndex();
   }
 
-  block(name:string, func:(block:DSLBlock) => any) {
-    let block = new DSLBlock(name, func, this);
+  block(name:string, func:BlockFunction) {
+    let block = new DSLBlock(name, func, this, undefined, undefined);
```
Comment (ibdknox) [inline src/runtime/dsl.ts]: don't need the undefineds here now

File: src/runtime/dsl.ts [hunk 4]
```diff
@@ -1129,7 +1188,7 @@ export class Program {
       }
     }
     trans.exec(this.index);
-    console.log(trans.changes.map((change, ix) => `    <- ${change}`).join("\n"));
+    // console.log(trans.changes.map((change, ix) => `    <- ${change}`).join("\n"));
```
Comment (ibdknox) [inline src/runtime/dsl.ts]: we should make this console.info

File: src/runtime/runtime.ts [hunk 5]
```diff
@@ -39,6 +39,12 @@ export function printConstraint(constraint:Constraint) {
   }
 }
 
+export function maybeReverse(value?:ID):ID|RawValue|undefined {
+  if(!value) return undefined;
```
Comment (ibdknox) [inline src/runtime/runtime.ts]: this should be `return value`

File: src/watchers/html.ts [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+  removeStyleInstance(styleId:string, instanceId:string) {
+    let instance = this.instances[instanceId];
+    if(!instance) return;
+    instance.removeAttribute("style");
+    let ix = instance.__styles!.indexOf(styleId);
+    instance.__styles!.splice(ix, 1);
+
+    for(let otherStyleId of instance.__styles!) {
+      console.log("reapplying", otherStyleId);
```
Comment (ibdknox) [inline src/watchers/html.ts]: 🗑

File: src/watchers/html.ts [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+  getRoot(id:string, tagname:RawValue = "div"):Instance {
+    return this.roots[id] = this.roots[id] || document.createElement(tagname as string);
+  }
+
+  clearRoot(id:string) {
+    this.roots[id] = undefined;
+  }
+
+  insertChild(parent:Instance, child:Instance) {
+    parent.appendChild(child);
```
Comment (ibdknox) [inline src/watchers/html.ts]: we need to sort these guys

File: src/watchers/html.ts [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+          style.add("tag", "html/style")
+        ];
+      })
+      .watch("Export html styles.", ({find, record, lib, not, lookup}) => {
+        let style = find("html/style");
+        let {attribute, value} = lookup(style);
+        return [
+          style.add(attribute, value)
+        ];
+      })
```
Comment (ibdknox) [inline src/watchers/html.ts]: we can probably collapse the block and watch here

File: src/runtime/dsl.ts (hunk 2 above)
Comment (joshuafcole) [inline src/runtime/dsl.ts]: J/k, this tracks the nodeIds now that they aren't all awesome nodes.

File: src/watchers/html.ts (hunk 8 above)
Comment (joshuafcole) [inline src/watchers/html.ts]: :+1:

**Comment by Eucalyptus2013** [comment]
Is it related to [this rfc](https://github.com/witheve/rfcs/pull/11) ?

**Pull request merged**